lib/pbench/server/utils.py

```diff
@@ -72,17 +73,18 @@ def quarantine(dest, logger, *files):
         sys.exit(101)
 
     for afile in files:
-        if not os.path.exists(afile) and not os.path.islink(afile):
```

Comment:
How is the original logic wrong?

If you have a two symbolic links, one that references a file that does exist, and one that references a file that does not exist, `os.path.exists(afile)` will return `True` for the symbolic link reference to the existing file, and will return `False` for the symbolic link reference to the non-existent file.

With that, this test is trying to make sure that we "quarantine" a symbolic link that references a non-existent file.

So I am not sure why this code needs to be changed.

Reply:
OK, I missed the subtlety that `os.path.exists` on a link tests for existence of the link destination; however this `continue` triggers only if `afile` isn't a link at all. So what we're actually saying here is that the path is a non-link file that doesn't exist. I still don't understand how this becomes a meaningful condition?

1. What do we want `quarantine` to do when given a path that's a real file (not a link)? This happens in put-shim, and is what drew my attention to this code.
2. What do we want `quarantine` to do when given a path that's a link to a file that exists? (We probably reasonably guess this to be a state link and rename it.)
3. What do we want `quarantine` to do when given a path that's a link to a file that doesn't exist, and how would this come about? (This might be, as your answer suggests, the intent of this code; but if so it seems it should be `and os.path.islink(afile)`.)

Reply:
> this `continue` triggers only if `afile` isn't a link at all.

Correct because we cannot invoke `shutil.move()` on a file that does not exist, but we can on a symbolic link that does exist, but links to a file that does not exist.

>  So what we're actually saying here is that the path is a non-link file that doesn't exist. I still don't understand how this becomes a meaningful condition?

We are just protecting the `shutil.move()` from the caller.

> 1. What do we want `quarantine` to do when given a path that's a real file (not a link)? This happens in put-shim, and is what drew my attention to this code.

Move it.

> 2. What do we want `quarantine` to do when given a path that's a link to a file that exists? (We probably reasonably guess this to be a state link and rename it.)

Move it.

> 3. What do we want `quarantine` to do when given a path that's a link to a file that doesn't exist,

Move it (in the best imitation of Sacha Baron Cohen as King Julien XIII possible).

> and how would this come about?

I don't know how, we just need to quarantine the unexpected condition.

> (This might be, as your answer suggests, the intent of this code; but if so it seems it should be `and os.path.islink(afile)`.)

Not sure I understand: We do not want to call `shutil.move()` on a file that does not exist.  So, a file *really* doesn't exist if it does not exist and it is not a symbolic link (thank you Python for being _so_ helpful!).  ;)

Reply:
Given how hard I am having to think about this, there is definitely **_something_** wrong here.  😁

This code skips `afile` if the path does not exist _and_ if the path does not refer to a symbolic link file.  So, if there were a symlink and it referred to a non-existent file, then the existence check would return `False`, and the `islink()` check would return `True`, and it would not be skipped, and it would be quarantined.

`exists()` | `islink()` | Quarantined
--- | --- | ---
False | True | Yes
True | True | Yes
True | False | Yes
False | False | No

Thus, in addition to the first case, we would also quarantine it if the symlink refers to an existent file as well as if it's an existing, real file (and not a symlink).

So, the only time we would skip the quarantine is when the file actually doesn't exist (i.e., it's not there in any form, symlink or otherwise)...which raises the question of how we know to look for it and why we're trying to prevent problems from touching it ([rather than recovering from them](https://docs.python.org/3.5/glossary.html#term-eafp)).

So, yeah, I think I concur that the original logic was appropriate.

The new logic will not quarantine existing but broken symlinks.  But, I have no idea what the ramifications of that are.

Reply:
OK; so, in summary, what's not clear here is the nature of the workaround to protect `shutil.move()`.  We really do want to quarantine any file, skipping only a file that __really__ doesn't exist instead of only __mostly__ doesn't exist. That is, `not os.path.exists()` isn't good enough here because we don't want to skip a broken symlink, and a symlink's `exists` is tied to the symlink's target rather than the symlink itself. The `and not os.path.islink` essentially removes the implicit symlink dereference from `exists` to avoid skipping the "mostly doesn't exist" case.

That's twisted.

Given this, I think the real resolution for the issue is to add that information to the `quarantine` function, either as part of the docstring or a comment. (I would prefer the docstring as it's generally more visible.)

Reply:
OK, nevermind...sorry for showing up late!  😆

Reply:
(The fact that my e-mail updates faster than the web page is a little bit...dizzying.  😝)

> (I would prefer the docstring as it's generally more visible.)

I would vote for "both":  explain in the docstring that we move all files, including symlink files; and, put a reminder in a code comment that `exists()` refers to the _target_ when the path refers to a symlink.  (In fact, I would be inclined to reverse the order of the `if` statement -- only check for (non-)existence if the path is _not_ to a symlink.)

Reply:
Let's close this PR, and the issue, and open a new issue for future work to address this.  If somebody wants to make a PR to do as Webb suggests, great.

Thanks all!

Reply:
I will close this PR